### PR TITLE
Port remaining caf_net tests to new framework

### PR DESCRIPTION
**Pull request by shariarriday**
Migrates remaining caf_net tests to updated test framework after https://github.com/actor-framework/actor-framework/pull/1786 is merged.

Remaining tests to migrate:
net.ssl.transport due to it's dependence on ssl certificates and pem.hpp

Relates https://github.com/actor-framework/actor-framework/issues/1731

### CODE REVIEW

File: libcaf_net/caf/net/multiplexer.test.cpp [hunk 1]
```diff
@@ -257,30 +290,32 @@ SCENARIO("a multiplexer terminates its thread after shutting down") {
...
 //           exhaust();
-//           CHECK_EQ(carl->name, "Carl");
-//           CHECK_EQ(carl->receive(), "Hello Carl!");
+//           check_eq(carl->name, "Carl");
+//           check_eq(carl->receive(), "Hello Carl!");
 //         }
 //       }
 //     }
 //   }
 // }
```
Comment (shariarriday) [inline libcaf_net/caf/net/multiplexer.test.cpp:317]: This is a commented out test which can be removed from the new tests. Are we planning to keep this test?

File: libcaf_net/caf/net/multiplexer.test.cpp [hunk 2]
```diff
@@ -24,19 +24,46 @@ namespace {
...
+    }
+  }
+
+private:
+  ptrdiff_t num_threads_;
+  std::mutex mx_;
+  std::atomic<ptrdiff_t> count_;
+  std::condition_variable cv_;
+};
+
```
Comment (Neverlord) [inline libcaf_net/caf/net/multiplexer.test.cpp:53]:
```suggestion
```

Please drop this and just use `caf::detail::latch` instead.

File: libcaf_net/caf/net/multiplexer.test.cpp (hunk 1 above)
Comment (Neverlord) [reply libcaf_net/caf/net/multiplexer.test.cpp:317]: It has a todo comment on it, so please leave it for now.

File: libcaf_net/caf/net/pipe_socket.test.cpp [hunk 3]
```diff
@@ -2,28 +2,35 @@
...
-CAF_TEST(send and receive) {
+namespace {
+
+template <class T>
+T unbox(caf::expected<T> x) {
+  if (!x)
+    test::runnable::current().fail("{}", to_string(x.error()));
+  return std::move(*x);
+}
+
```
Comment (Neverlord) [inline libcaf_net/caf/net/pipe_socket.test.cpp:22]:
```suggestion
```

This is used just once in this line:

```c++
std::tie(rd_sock, wr_sock) = unbox(make_pipe());
```

We can just replace that with:

```
auto maybe_sockets = make_pipe();
require(maybe_sockets.has_value());
std::tie(rd_sock, wr_sock) = *maybe_sockets;
```

File: libcaf_net/caf/net/stream_socket.test.cpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        detail::format("cannot create socket with error {}", err).c_str());
+    if (auto err = nonblocking(second, true); err != caf::none)
+      CAF_RAISE_ERROR(
+        detail::format("cannot create socket with error {}", err).c_str());
+    if (auto buffer_size = unbox(send_buffer_size(first)); buffer_size == 0u)
+      CAF_RAISE_ERROR(
+        detail::format("buffer size for first: {}", buffer_size).c_str());
+    if (auto buffer_size = unbox(send_buffer_size(second)); buffer_size == 0u)
+      CAF_RAISE_ERROR(
+        detail::format("buffer size for first: {}", buffer_size).c_str());
```
Comment (Neverlord) [inline libcaf_net/caf/net/stream_socket.test.cpp:47]:
```suggestion
    require_eq(nonblocking(first, true), error{});
    require_eq(nonblocking(second, true), error{});
    require_eq(send_buffer_size(first), 0u);
    require_eq(send_buffer_size(second), 0u);
```

File: libcaf_net/caf/net/stream_socket.test.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+namespace {
+
+template <class T>
+T unbox(caf::expected<T> x) {
+  if (!x)
+    test::runnable::current().fail("{}", to_string(x.error()));
+  return std::move(*x);
+}
+
```
Comment (Neverlord) [inline libcaf_net/caf/net/stream_socket.test.cpp:24]:
```suggestion
```

Unnecessary (see other comment).

File: libcaf_net/caf/net/udp_datagram_socket.test.cpp [hunk 6]
```diff
@@ -59,30 +68,30 @@ error read_from_socket(udp_datagram_socket sock, byte_buffer& buf) {
...
 
-CAF_TEST(socket creation) {
+TEST("socket creation") {
   ip_endpoint ep;
-  CHECK_EQ(parse("0.0.0.0:0", ep), none);
-  CHECK(make_udp_datagram_socket(ep));
-}
+  check_eq(parse("0.0.0.0:0", ep), none);
+  check(static_cast<bool>(make_udp_datagram_socket(ep)));
+} // WITH_FIXTURE(fixture)
```
Comment (Neverlord) [inline libcaf_net/caf/net/udp_datagram_socket.test.cpp:77]: I don't think this belongs here?

```suggestion
}
```

File: libcaf_net/caf/net/stream_socket.test.cpp (hunk 4 above)
Comment (shariarriday) [reply libcaf_net/caf/net/stream_socket.test.cpp]: We can not currently use the runnable instance before the fixture is constructed. Due to this, the `require_eq()` throws error and the test fails directly.

File: libcaf_net/caf/net/stream_socket.test.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    if (auto err = nonblocking(second, true); err != caf::none)
+      CAF_RAISE_ERROR(
+        detail::format("cannot create socket with error {}", err).c_str());
+    if (auto buffer_size = send_buffer_size(first); buffer_size == 0u)
+      CAF_RAISE_ERROR(
+        detail::format("buffer size for first: {}", buffer_size).c_str());
+    if (auto buffer_size = send_buffer_size(second); buffer_size == 0u)
+      CAF_RAISE_ERROR(
+        detail::format("buffer size for first: {}", buffer_size).c_str());
+  }
```
Comment (Neverlord) [inline libcaf_net/caf/net/stream_socket.test.cpp:44]:
```suggestion
    auto maybe_sockets = make_stream_socket_pair();
    if (!maybe_sockets)
      CAF_RAISE_ERROR("cannot create connected socket pair");
    std::tie(first, second) = *maybe_sockets;
    if (auto err = nonblocking(first, true))
      CAF_RAISE_ERROR(
        detail::format("failed to set socket to nonblocking: {}", err).c_str());
    if (auto err = nonblocking(second, true))
      CAF_RAISE_ERROR(
        detail::format("failed to set socket to nonblocking: {}", err).c_str());
    if (auto buffer_size = send_buffer_size(first); buffer_size == 0u)
      CAF_RAISE_ERROR("failed to set send_buffer_size");
    if (auto buffer_size = send_buffer_size(second); buffer_size == 0u)
      CAF_RAISE_ERROR("failed to set send_buffer_size");
  }
```

File: libcaf_net/caf/net/udp_datagram_socket.test.cpp [hunk 8]
```diff
@@ -2,30 +2,39 @@
...
 using namespace caf;
 using namespace caf::net;
 using namespace caf::net::ip;
 
 namespace {
 
+template <class T>
+T unbox(caf::expected<T> x) {
+  if (!x)
+    CAF_RAISE_ERROR(detail::format("{}", to_string(x.error())).c_str());
```
Comment (Neverlord) [inline libcaf_net/caf/net/udp_datagram_socket.test.cpp:27]:
```suggestion
    CAF_RAISE_ERROR(to_string(x.error()).c_str());
```

**Review by Neverlord — CHANGES_REQUESTED** [review]

**Pull request merged**